driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java

```diff
@@ -126,23 +127,43 @@ public ServerDescription getDescription() {
 
     @Override
     public void invalidate() {
+        invalidate(connectionPool.getGeneration());
+    }
+
+    @Override
+    public void invalidate(final int connectionGeneration) {
         if (!isClosed()) {
+            if (connectionGeneration < connectionPool.getGeneration()) {
+                return;
+            }
             serverStateListener.stateChanged(new ChangeEvent<ServerDescription>(description, ServerDescription.builder()
                     .state(CONNECTING)
                     .address(serverId.getAddress())
                     .build()));
             connectionPool.invalidate();
+            // TODO: in streaming protocol, we want to close the current connection and start over
             connect();
         }
     }
 
     @Override
-    public void invalidate(final Throwable t) {
+    public void invalidate(final Throwable reason) {
+        invalidate(reason, connectionPool.getGeneration(), description.getMaxWireVersion());
+    }
+
+    @Override
+    public void invalidate(final Throwable t, final int connectionGeneration, final int maxWireVersion) {
         if (!isClosed()) {
+            if (connectionGeneration < connectionPool.getGeneration()) {
```

Comment:
First requirement again, but in the invalidate method used for already-opened connections, where we also have access to the wire version